File: computational_physics/Homework2/hw_2_2.jl
function gauss_elimination(A::Matrix, b::Vector)::Vector
    n = length(b)
    Ab = [A b]# stacking the augmented matrix and the vector
    
    # Forward elimination
    for k in 1:n
        # Finding the pivot row and swapping rows if required ; although pivoting is not required for this particular problem
        maxval, pivotrow = findmax(abs.(Ab[k:n, k]))
        pivotrow += k - 1
        Ab[k, :], Ab[pivotrow, :] = Ab[pivotrow, :], Ab[k, :]
        # display(Ab)
        # Eliminate the entries below the pivot
        for i in k+1:n
            factor = Ab[i, k] / Ab[k, k]
            # println(factor)
            # display(Ab[i,k:n+1])
            # display(Ab[k,k:n+1])
            # display(factor .* Ab[k, k:n+1])
            Ab[i, k:n+1] = Ab[i, k:n+1].-factor .* Ab[k, k:n+1]
        end
    end
    
    # Backward substitution
    x = zeros(Float64, n)
    for i in n:-1:1
        x[i] = (Ab[i, end] - sum(Ab[i, i+1:n].* x[i+1:n])) / Ab[i, i]
    end
    
    return x
end

# Coefficient matrix A 
A = [10.0 -4.0 3.0 2.0; 1.0 -1.0 -2.0 1.0; 3.0 2.0 -1.0 2.0; 2.0 3.0 1.0 -3.0]
display(A)
b = [3.0, 1.0, 4.0, 2.0]
display(b)

# Solving the system of equations
@time x = gauss_elimination(A, b)
println("Solution:")
println("x = ", x[1])
println("y = ", x[2])
println("z = ", x[3])
println("u = ", x[4])

File: computational_physics/Homework2/hw_2_2.cpp
#include <iostream>
#include <vector>

// Function for Gaussian elimination
std::vector<double> gauss_elimination(std::vector<std::vector<double>> A, std::vector<double> b) {
    int n = b.size();
    std::vector<std::vector<double>> Ab(n, std::vector<double>(n + 1));

    // Initialize the augmented matrix Ab
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            Ab[i][j] = A[i][j];
        }
        Ab[i][n] = b[i];
    }

    // Forward elimination
    for (int k = 0; k < n; ++k) {
        // Finding the pivot row and swapping rows if required (pivoting)
        int pivotrow = k;
        for (int i = k + 1; i < n; ++i) {
            if (std::abs(Ab[i][k]) > std::abs(Ab[pivotrow][k])) {
                pivotrow = i;
            }
        }
        std::swap(Ab[k], Ab[pivotrow]);

        // Eliminate the entries below the pivot
        for (int i = k + 1; i < n; ++i) {
            double factor = Ab[i][k] / Ab[k][k];
            for (int j = k; j <= n; ++j) {
                Ab[i][j] -= factor * Ab[k][j];
            }
        }
    }

    // Backward substitution
    std::vector<double> x(n);
    for (int i = n - 1; i >= 0; --i) {
        x[i] = Ab[i][n];
        for (int j = i + 1; j < n; ++j) {
            x[i] -= Ab[i][j] * x[j];
        }
        x[i] /= Ab[i][i];
    }

    return x;
}

int main() {
    // Coefficient matrix A
    std::vector<std::vector<double>> A = {
        {10.0, -4.0, 3.0, 2.0},
        {1.0, -1.0, -2.0, 1.0},
        {3.0, 2.0, -1.0, 2.0},
        {2.0, 3.0, 1.0, -3.0}
    };

    // Right-hand side vector b
    std::vector<double> b = {3.0, 1.0, 4.0, 2.0};

    // Solving the system of equations using Gaussian elimination
    std::vector<double> x = gauss_elimination(A, b);

    // Printing the solution
    std::cout << "Solution:" << std::endl;
    for (size_t i = 0; i < x.size(); ++i) {
        std::cout << "x[" << i << "] = " << x[i] << std::endl;
    }

    return 0;
}
